projects/vehicles/plugins/robot_windows/automobile/automobile.c

```diff
@@ -0,0 +1,397 @@
+#include <webots/robot.h>
+#include <webots/robot_wwi.h>
+#include <webots/utils/default_robot_window.h>
+
+#include <webots/vehicle/car.h>
+#include <webots/vehicle/driver.h>
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <sys/time.h>
+#include <unistd.h>
+#endif
+
+#include "generic.h"
+#include "string_utils.h"
+
+enum DriverInfo {
+  NONE = 0,
+  OVERVIEW_INFO = 1,
+  SPEED_INFO,
+  STEERING_INFO,
+  ENCODERS_INFO,
+  BRAKE_INFO,
+  THROTTLE_INFO,
+  RPM_INFO,
+  COUNT
+};
+static bool driverInfoEnabled[COUNT];
+
+static char *buffer = NULL;
+static int buffer_size = 0;
+
+static void free_buffer() {
+  free(buffer);
+  buffer = NULL;
+}
+
+static void create_buffer(int size) {
+  atexit(free_buffer);
+  buffer_size = size + 1;
+  buffer = malloc(buffer_size);
+}
+
+static void buffer_append(const char *string) {
+  if (string == NULL || string[0] == '\0')
+    return;
+  const int l = strlen(string);
+  if (buffer == NULL) {
+    create_buffer(l);
+    memcpy(buffer, string, buffer_size);
+    return;
+  }
+  buffer = realloc(buffer, buffer_size + l);
+  if (buffer == NULL) {
+    fprintf(stderr, "Error creating message to be sent to the robot window: not enough memory.\n");
+    exit(EXIT_FAILURE);
+  }
+  memcpy(&buffer[buffer_size - 1], string, l + 1);
+  buffer_size += l;
+}
+
+static long int last_update_time = 0;
+
+static long int current_time() {
+#ifdef _WIN32
+  SYSTEMTIME tim;
+  GetSystemTime(&tim);
+  return (((tim.wHour * 60 + tim.wMinute) * 60 + tim.wSecond) * 1000 + tim.wMilliseconds) + 0.5;
+#else
+  struct timeval tim;
+  gettimeofday(&tim, NULL);
+  return ((tim.tv_sec) * 1000 + tim.tv_usec / 1000.0) + 0.5;
+#endif
+}
+
+static enum DriverInfo driver_info_from_string(const char *s) {
+  if (strcmp(s, "Overview") == 0)
+    return OVERVIEW_INFO;
+  if (strcmp(s, "Speed") == 0)
+    return SPEED_INFO;
+  if (strcmp(s, "Steering") == 0)
+    return STEERING_INFO;
+  if (strcmp(s, "Encoders") == 0)
+    return ENCODERS_INFO;
+  if (strcmp(s, "Brake") == 0)
+    return BRAKE_INFO;
+  if (strcmp(s, "Throttle") == 0)
+    return THROTTLE_INFO;
+  if (strcmp(s, "RPM") == 0)
+    return RPM_INFO;
+  return NONE;
+}
+
+static void enable_driver_info(enum DriverInfo info, bool enable) {
+  if (info <= NONE || info >= COUNT) {
+    assert(0);
+    return;
+  }
+  driverInfoEnabled[info] = enable;
+}
+
+// JavaScript -> C protocol description:
+//   [deviceName:commandTag[=commadState][,]]*
+// example:
+//   "e-puck:forward,ds0:enable,myMotor0:value=1.2"
+static void apply_command(const char *command) {
+  char *tokens = strdup(command);
+  char *token = NULL;
+
+  enum DriverInfo info = NONE;
+  WbDeviceTag tag = 0;
+  bool robot = false;
+
+  while ((token = string_utils_strsep(&tokens, ":"))) {
+    if (info == NONE && tag == 0 && !robot) {  // first token = device or robot name
+      char *name0 = string_utils_replace(token, "\\:", ":");
```

Comment:
`const`?

Reply:
I will address this first in #2655

Reply:
I checked but this variable cannot be turned to const because the `free` function would discard the `const` qualifier.